File: 244_ShortestWordDistanceII.java
public class WordDistance {

    Map<String, List<Integer>> positions;
    public WordDistance(String[] words) {
        positions = new HashMap<>();
        for (int i=0; i<words.length; i++) {
            String key = words[i];
            if (!positions.containsKey(key)) {
                List<Integer> lst = new ArrayList<>();
                lst.add(i);
                positions.put(key, lst);
            } else {
                positions.get(key).add(i);
            }
        }
    }
    
    public int shortest(String word1, String word2) {
        List<Integer> list1 = positions.get(word1);
        List<Integer> list2 = positions.get(word2);
        int dis = Integer.MAX_VALUE;
        int m = list1.size(), n = list2.size();
        int i = 0, j = 0;
        while (i < m && j < n) {
            while (i < m && list1.get(i) < list2.get(j)) i++;
            if (i > 0) dis = Math.min(dis, Math.abs(list1.get(i-1) - list2.get(j)));
            if (i == m) break;
            while (j < n && list2.get(j) < list1.get(i)) j++;
            if (j > 0) dis = Math.min(dis, Math.abs(list1.get(i) - list2.get(j-1)));
        }
        return dis;
    }
}

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance obj = new WordDistance(words);
 * int param_1 = obj.shortest(word1,word2);
 */

File: 244_ShortestWordDistanceII.py
class WordDistance(object):

    def __init__(self, words):
        """
        :type words: List[str]
        """
        self.dict = {}
        for i in range(len(words)):
            if words[i] in self.dict:
                self.dict[words[i]].append(i)
            else:
                self.dict[words[i]] = [i]



    def shortest(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        list1 = self.dict[word1]
        list2 = self.dict[word2]
        i, j, dis = 0, 0, abs(list1[0] - list2[0])
        while i < len(list1) and j < len(list2):
            dis = min(dis, abs(list1[i] - list2[j]))
            if list1[i] < list2[j]:
                i += 1
            else:
                j += 1
        return dis


# Your WordDistance object will be instantiated and called as such:
# obj = WordDistance(words)
# param_1 = obj.shortest(word1,word2)

class WordDistance(object):
    def __init__(self, words):
        self.d = {}
        for i, w in enumerate(words):
            self.d[w] = self.d.get(w, []) + [i]

    def shortest(self, w1, w2):
        a, b = self.d[w1], self.d[w2]
        m, n, i, j, res = len(a), len(b), 0, 0, sys.maxsize
        while i < m and j < n:
            res = min(res, abs(a[i] - b[j]))
            if a[i] < b[j]:
                i += 1
            else:
                j += 1
        return res